### DISCUSSION TITLE
Vite 2.x 下如何使用 antd ?

### DISCUSSION THREAD

**Discussion by yunsii** [ant-design/ant-design#29660] [Show and tell]
**答**：基于 `vite-plugin-imp`，支持定制主题。

```ts
import vitePluginImp from 'vite-plugin-imp';

export default defineConfig({
  plugins: [
    vitePluginImp({
      libList: [
        {
          libName: 'antd',
          style: (name) => `antd/es/${name}/style`,
        },
      ],
    }),
  ],
  css: {
    preprocessorOptions: {
      less: {
        // modifyVars: { 'primary-color': '#13c2c2' },
        javascriptEnabled: true,
      },
    },
  },
});
```

更多配置可以参考我项目里的配置 [vite-react/vite.config.ts](https://github.com/theprimone/vite-react/blob/master/vite.config.ts)。至于讨论 https://github.com/ant-design/ant-design/discussions/29325 中提到组件无法使用的情况暂时没遇到，遇到就 PR \_(:3J∠)\_

**Comment by yunsii** [comment]
~果然发现了一个问题，使用 List 组件 loading 效果不可见，需要单独引入 spin 的样式。~

问题已解决，配置已更新。

**Reply by yunsii** [reply]
看了 https://github.com/vitejs/vite/issues/728#issuecomment-760260219 才反应过来，有点搞笑了。没注意直接用的 lib，然后报错 `require is not defined`，就想着直接引用 less，看到突然恍然大悟，用 es 就完事了，样式依赖的问题就不用自己搞了。

**Comment by yunsii** [comment]
今天又发现了一个问题，当在 Vite 下混用 antd/es 和 antd/lib 时，编译会有问题。原因是 @babel/runtime 在 antd/lib 和其内部使用方式不同：

在 [antd/lib/locale-provider/index.js](https://unpkg.com/browse/antd@4.17.4/lib/locale-provider/index.js) 中：

```js
var _typeof = require("@babel/runtime/helpers/typeof");
```

在 [@babel/runtime/helpers/interopRequireWildcard.js](https://unpkg.com/browse/@babel/runtime@7.16.5/helpers/interopRequireWildcard.js) 中：

```js
var _typeof = require("@babel/runtime/helpers/typeof")["default"];
```

因此，以我当前的认知来看，无论 vite 中 commonjsOption 如何配置编译 `require`，都不能调和这两个矛盾。将 antd/lib 引用全部换成 antd/es 编译问题得以解决。最后，由此可得，在 vite 中开发时，能用 esm 就用 esm，别乱混用 cjs 的东西 :joy:

**Reply by yoyo837** [reply]
分享下配置呗。

**Reply by yunsii** [reply]
> 分享下配置呗。

没别的配置了啊 :joy: 这里提到的是我碰到的 cjs 和 esm 混用的问题，都用 esm 就好了。

**Reply by yunsii** [reply]
> 分享下配置呗。

是我表达的问题吗？现在更新了一下，一楼就是相关配置 \_(:з」∠)\_

**Comment by Emily-0616** [comment]
我根据这份配置出现了另外一个问题，浏览器log 输出的行数与实际开发报错的行数不一致，我修改了sourcemap 同样没效果

**Reply by yunsii** [reply]
我也碰到了，不过是后来发现的，不确定是什么插件导致的问题，后来搁置了，刚去 https://github.com/onebay/vite-plugin-imp/issues 看了一下，看起来问题不小。

如果问题不好解决的情况下，可以考虑直接使用 `antd/dist/antd.less` 而不用按需加载样式。

**Comment by hxing** [comment]
------------------ 原始邮件 ------------------ 
    
   发件人："云深" ***@***.***" 
   时　间：2022/07/06 11:45:37 周三 
   收件人：ant-designant-design ***@***.***" 
   抄送人：Subscribed ***@***.***" 
   主　题：Re: [ant-design/ant-design] Vite 2.x 下如何使用 antd ? (Discussion #29660) 
    
    
     
   我也碰到了，不过是后来发现的，不确定是什么插件导致的问题，后来搁置了，刚去 https://github.com/onebay/vite-plugin-imp/issues 看了一下，看起来问题不小。 
   如果问题不好解决的情况下，可以考虑直接使用 antd/dist/antd.less 而不用按需加载样式。 
   —Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: 
      
       @
       github
       .
       com>
